compiler/rustc_passes/src/liveness.rs

```diff
@@ -188,6 +188,19 @@ enum VarKind {
     Upvar(HirId, Symbol),
 }
 
+struct CollectLitsVisitor<'tcx> {
+    lit_exprs: Vec<&'tcx hir::Expr<'tcx>>,
+}
+
+impl<'tcx> Visitor<'tcx> for CollectLitsVisitor<'tcx> {
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        if let hir::ExprKind::Lit(_) = expr.kind {
+            self.lit_exprs.push(expr);
```

Comment:
instead of pushing these on a vector, you could just report the lint here directly.

Reply:
If we report the lint here directly, we will have to pass the `err `from `let mut err = lint.build(&format!("unused variable: `{}`", name))`  to the visitor, also we will need to pass the `name `from `"{somevar}"` and a tag to tell if we found those strings. That makes things more complicated. 
Also, the compiler seems don't like me to pass the `err `from `let mut err = lint.build(&format!("unused variable: `{}`", name))`  to the visitor and complains the lifetime of it:
```
error[E0597]: `err` does not live long enough
    --> compiler/rustc_passes/src/liveness.rs:1720:76
     |
1720 |                                 let mut visitor = CollectLitsVisitor {verr:&mut err, name:&name, emit_string_interpolation:false};
     |                                                                            ^^^^^^^^ borrowed value does not live long enough
...
1733 |                         },
     |                         -
     |                         |
     |                         `err` dropped here while still borrowed
     |                         borrow might be used here, when `err` is dropped and runs the destructor for type `DiagnosticBuilder<'_, ()>`

```